server/block/hopper.go

```diff
@@ -0,0 +1,269 @@
+	}
+}
+
+// Model ...
+func (Hopper) Model() world.BlockModel {
+	return model.Hopper{}
+}
+
+// SideClosed ...
+func (Hopper) SideClosed(cube.Pos, cube.Pos, *world.World) bool {
+	return false
+}
+
+// BreakInfo ...
+func (h Hopper) BreakInfo() BreakInfo {
+	return newBreakInfo(3, pickaxeHarvestable, pickaxeEffective, oneOf(h)).withBlastResistance(24)
+}
+
+// Inventory returns the inventory of the hopper.
+func (h Hopper) Inventory(w *world.World, pos cube.Pos) *inventory.Inventory {
+	return h.inventory
+}
+
+// WithName returns the hopper after applying a specific name to the block.
+func (h Hopper) WithName(a ...any) world.Item {
+	h.CustomName = strings.TrimSuffix(fmt.Sprintln(a...), "\n")
+	return h
+}
+
+// AddViewer adds a viewer to the hopper, so that it is updated whenever the inventory of the hopper is changed.
+func (h Hopper) AddViewer(v ContainerViewer, _ *world.World, _ cube.Pos) {
+	h.viewerMu.Lock()
+	defer h.viewerMu.Unlock()
+	h.viewers[v] = struct{}{}
+}
+
+// RemoveViewer removes a viewer from the hopper, so that slot updates in the inventory are no longer sent to it.
+func (h Hopper) RemoveViewer(v ContainerViewer, _ *world.World, _ cube.Pos) {
+	h.viewerMu.Lock()
+	defer h.viewerMu.Unlock()
+	delete(h.viewers, v)
+}
+
+// Activate ...
+func (Hopper) Activate(pos cube.Pos, _ cube.Face, _ *world.World, u item.User, _ *item.UseContext) bool {
+	if opener, ok := u.(ContainerOpener); ok {
+		opener.OpenBlockContainer(pos)
+		return true
+	}
+	return false
+}
+
+// UseOnBlock ...
+func (h Hopper) UseOnBlock(pos cube.Pos, face cube.Face, _ mgl64.Vec3, w *world.World, user item.User, ctx *item.UseContext) bool {
+	pos, _, used := firstReplaceable(w, pos, face, h)
+	if !used {
+		return false
+	}
+
+	//noinspection GoAssignmentToReceiver
+	h = NewHopper()
+	h.Facing = cube.FaceDown
+	if h.Facing != face {
+		h.Facing = face.Opposite()
+	}
+
+	place(w, pos, h, user, ctx)
+	return placed(ctx)
+}
+
+// Tick ...
+func (h Hopper) Tick(currentTick int64, pos cube.Pos, w *world.World) {
+	h.TransferCooldown--
+	h.CollectCooldown--
+	h.LastTick = currentTick
+
+	if h.TransferCooldown >= 0 || h.CollectCooldown >= 0 {
+		w.SetBlock(pos, h, nil)
+		return
+	}
+
+	h.TransferCooldown = 0
+	h.CollectCooldown = 0
+	if h.Powered {
+		w.SetBlock(pos, h, nil)
+		return
+	}
+
+	inserted := h.insertItem(pos, w)
+	extracted := h.extractItem(pos, w)
+	if inserted || extracted {
+		h.TransferCooldown = 8
+		w.SetBlock(pos, h, nil)
+	}
+}
+
+// HopperInsertable represents a block that can have its contents inserted into by a hopper.
+type HopperInsertable interface {
+	// InsertItem handles the insert logic for that block.
+	InsertItem(h Hopper, pos cube.Pos, w *world.World) bool
+}
+
+// insertItem inserts an item into a block that can receive contents from the hopper.
+func (h Hopper) insertItem(pos cube.Pos, w *world.World) bool {
+	dest := w.Block(pos.Side(h.Facing))
+	if e, ok := dest.(HopperInsertable); ok {
+		return e.InsertItem(h, pos, w)
```

Comment:
Why not pass the `dest` position so that every `HopperInsertable` implementer does not have to write `pos.Face(h.Facing)`?